# mk_tests

Scripts to implement an McDonald-Kreitman test from a set of genome 
sequence, annotations, and variants.

## General overview

The script requires an annotated genome, and variants for both a group
of focal individuals and an outgroup. Using this:

1. Obtain all the per-sample coding sequences (with genotyped variants)
for all individuals (`extract_hap_cds.py`)
2. Use `orthofinder` to cluster all the coding-sequence and finding single-
copy orthologs for focal and outgroup individuals (script TDB).
3. Perform a codon-aware re-alignment of the single-copy orthologs
inferred from `orthofinder` using `macse` (script TBD).
4. For each single-copy ortholog MSA, process the sequences and run a custom
McDoland-Kreitman test (`run_ruby_mkt_from_msa.py`).

## Python script to extract the CDSs

The program `extract_hap_cds.py` extracts the per-sample coding sequences using 
`samtools faidx` and `bcftools consensus`. The commands process each individual 
CDS entry in the GFF individually, and merges them into the corresponding 
transcripts when necessary.

```sh
$ python3 extract_hap_cds.py -h
  usage: extract_hap_cds.py [-h] -g GENOME -f GFF -v VCF [-o OUT_DIR] 
                            [-t THREADS] [--snps-only]

  options:
    -h, --help            show this help message and exit
    -g, --genome GENOME   (str) Path to genome in FASTA format.
    -f, --gff GFF         (str) Path to the annotation in FASTA format.
    -v, --vcf VCF         (str) Path to variants in VCF/BCF format.
    -o, --out-dir OUT_DIR
                          (str) Path to output directory [default=.].
    -t, --threads THREADS
                          (int) Number of threads to run in parallel sections 
                          of code [default=1].
    --snps-only           Filter the input variants to only keep SNPs.
```

## Calculate the codong positions of each coding site

The program `position_in_codon.py` parses an annotation file (GFF) and for 
each coding site, calculate the the codon number (i.e., which codon in amino 
acid sequence the site belongs to) and codon position (i.e., 1st, 2nd, or 3rd 
position of the codon).

```sh
$ python3 position_in_codon.py -h
  position_in_codon.py started on 2025-05-08 23:47:18.
  usage: position_in_codon.py [-h] -g GFF [-o OUT_DIR]

  Parse an annotation file (GFF) and for each coding site, calculate the the codon
  number (i.e., which codon in amino acid sequence the site belongs to) and codon
  position (i.e., 1st, 2nd, or 3rd position of the codon).
  
  options:
    -h, --help            show this help message and exit
    -g GFF, --gff GFF     (str) Path to the annotation in GFF format.
    -o OUT_DIR, --out-dir OUT_DIR
                          (str) Path to output directory [default=.].
```

The script generates the output table, `position_in_codons.tsv`. 
Output example:

```sh
#Chrom  Bp     Transcript  CDS    CodonNum  PosInCodon
chr01   53103  mrna-1      cds-1  1         1
chr01   53104  mrna-1      cds-1  1         2
chr01   53105  mrna-1      cds-1  1         3
chr01   53106  mrna-1      cds-1  2         1
chr01   53107  mrna-1      cds-1  2         2
chr01   53108  mrna-1      cds-1  2         3
chr01   53109  mrna-1      cds-1  3         1
chr01   53110  mrna-1      cds-1  3         2
chr01   53111  mrna-1      cds-1  3         3
...
```

## Run `orthofinder`

TODO:

## Run `macse`

TODO 

## Script to automate the mkTest

The script `run_ruby_mkt_from_msa.py` takes all the processed alignments,
generated by clustering and aligning all the sequences generated by
`extract_hap_cds.py` (more details TBD), and run a custom ruby script to
perform a McDonald-Kreitman test.

The source of the ruby script is <https://github.com/andrewkern/mkTest>,
and it is a required dependency.

TODO: Describe options in more detail.

```sh
$ python3 run_ruby_mkt_from_msa.py -h
  run_ruby_mkt_from_msa.py started on 2025-05-12 21:01:15.
  usage: run_ruby_mkt_from_msa.py [-h] -l SCO_LIST -m MSA_DIR -g OUTGROUP_ID [-e EXE_DIR] [-o OUT_DIR]

  Run a custom Ruby script for MK-Test anslysis.

  options:
    -h, --help            show this help message and exit
    -l SCO_LIST, --sco-list SCO_LIST
                          (str) Path to the file containing list of single-copy orthologs.
    -m MSA_DIR, --msa-dir MSA_DIR
                          (str) Path to directory containing MSA files.
    -g OUTGROUP_ID, --outgroup-id OUTGROUP_ID
                          (str) ID of the outgroup species in the MSA.
    -e EXE_DIR, --exe-dir EXE_DIR
                          (str) Path to the Ruby `mkTest.rb` executable [default=.].
    -o OUT_DIR, --out-dir OUT_DIR
                          (str) Path to output directory [default=.].
```

### Output example

```sh
orthologID     focalGene   alnLen  avgAlnLenNoGaps  sdAlnLenNoGaps  aaFix  aaPoly  silFix  silPoly  fetPval  notes
N0.HOG0000111  mrna-18589  789     685.400          17.041          139    0       70      0        1        passed
N0.HOG0000139  mrna-17553  909     859.267          192.063         -1     -1      -1      -1       -1       frameshifts
N0.HOG0000208  mrna-8936   1332    1257.000         186.121         -1     -1      -1      -1       -1       frameshifts
N0.HOG0000209  mrna-16468  1356    1313.867         159.328         -1     -1      -1      -1       -1       frameshifts
N0.HOG0000231  mrna-18370  1131    1131.000         0.000           0      0       20      30       1        passed
N0.HOG0000323  mrna-17302  1065    995.200          16.606          209    64      134     41       1        passed
N0.HOG0000328  mrna-7643   651     551.600          29.890          52     15      32      21       0.0469   passed
N0.HOG0000482  mrna-17932  780     749.000          120.062         -1     -1      -1      -1       -1       frameshifts
N0.HOG0000489  mrna-13195  822     781.533          17.025          -1     -1      -1      -1       -1       frameshifts
```

TODO: Add details on columns.

## Authors

Angel G. Rivera-Colon  
Institute of Ecology and Evolution  
University of Oregon
